### PR TITLE
Completed current stage of Optimizer

### PR DESCRIPTION
**Pull request by choult**
* Added InsertDirectly class
 * Added FilterBeforeSort class

All tests passing

### CODE REVIEW

File: src/Optimizer/InsertDirectly.php [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+ * Rule: "insert(f(a, b), c) => insert(a, c)" when "f" is the "sort", "slice", or "filter" function
+ *   Note: this rule may apply more than once during the simplification of a request
+ *   Note: if this rule doesn't apply, then the request should be left unchanged
+*/
+class InsertDirectly
+{
+
+    public function optimize($token)
+    {
+        if (!(isset($token[1]) && $token[1] == 'insert')) {
```
Comment (smortensen) [inline src/Optimizer/InsertDirectly.php]: Let's assume that the user-supplied input may be invalid, but we _can_ trust the output from our code that we control completely.

So, the structure of the token will be guaranteed to be valid (and we won't need to check for its validity with an "isset"). But we can't assume that the user will have provided a valid insert query, or that the user-supplied arguments will match our specifications.

If the user hasn't supplied a first-argument to one of the "insert," "sort," "slice," or "filter" functions, then we should return the token without processing. We should treat this as a normal event: the code should work without triggering any PHP errors, warnings, or notices, and we shouldn't throw an exception.

We'll do in-depth query validation later on in the processing, and that logic will be the "single source of truth" for query validation.

File: src/Optimizer/InsertDirectly.php [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+
+use Datto\Cinnabari\Parser;
+
+/**
+ * An optimizer to collapse sort, slice and filter functions when the request method is "insert"
+ *
+ * Rule: "insert(f(a, b), c) => insert(a, c)" when "f" is the "sort", "slice", or "filter" function
+ *   Note: this rule may apply more than once during the simplification of a request
+ *   Note: if this rule doesn't apply, then the request should be left unchanged
+*/
```
Comment (smortensen) [inline src/Optimizer/InsertDirectly.php:35]: Let me add a comment, which may help clarify the problem:

There may be more than "insert" function in the query, and they might not appear at the top level of the query. For example, one or more "insert" functions might appear as a subexpression within a group or function token.

We should start processing at the top-level token in the request--preserving everything--until we arrive at an "insert" function: Then we should strip away any immediate-child sequence of "sort," "slice," or "filter" functions that appear in the first-argument slot. When we reach a different token (that isn't one of these three function tokens), then we should return that token without any processing.

File: src/Optimizer/FilterBeforeSort.php [hunk 3]
```diff
@@ -0,0 +1,77 @@
...
+    public function optimize($token)
+    {
+        return $this->convert($token);
+    }
+
+    private function convert($token)
+    {
+        if ($token[0] !== Parser::TYPE_FUNCTION) {
+            return $token;
+        }
```
Comment (smortensen) [inline src/Optimizer/FilterBeforeSort.php:67]: A group token will also have subexpressions, which may contain "filter-sort" instances too.

File: src/Optimizer/FilterBeforeSort.php [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+
+        $token[2] = $this->convert($token[2]);
```
Comment (smortensen) [inline src/Optimizer/FilterBeforeSort.php]: The user might not supply any arguments to a function. We should treat this as a normal event: a function without argument won't match our rule, so we'll return the token without processing.

File: src/Optimizer/FilterBeforeSort.php [hunk 5]
```diff
@@ -0,0 +1,77 @@
...
+                $token[1], // Should be 'filter'
+                $token[2][1], // Should be 'sort'
+                $token[3], // The filter's second argument
+                $token[2][3] // The sort's second argument
+            ) = array(
+                'sort',
+                'filter',
+                $token[2][3],   // The sort's second argument
+                $token[3]   // The filter's second argument
+            );
```
Comment (smortensen) [inline src/Optimizer/FilterBeforeSort.php:95]: The user might provide a different number of arguments. This shouldn't cause our code to fail with any PHP notices, warnings, or errors.

**Review by smortensen — CHANGES_REQUESTED** [review]
I've added comments to help explain the behavior of these two optimization rules. Because this behavior wasn't in the original notes or unit tests, I've added new units as well. You can get those new unit tests if you merge in the latest changes from the "Optimizer" branch.

Thank you,
~Spencer

File: src/Optimizer/InsertDirectly.php (hunk 1 above)
Comment (choult) [reply src/Optimizer/InsertDirectly.php]: Should the query not be optimized *after* validation? When collapsing the AST we should probably be reasonably secure in it being syntactically valid.

**Comment by choult** [comment]
All tests passing once more; I believe I've covered your points of feedback, but I have the nagging feeling that there are some edge-cases I'm not aware of/haven't defended against in the AST.

File: src/Optimizer/InsertDirectly.php (hunk 1 above)
Comment (smortensen) [reply src/Optimizer/InsertDirectly.php]: At this point, we are certain that the abstract syntax tree is syntactically valid: That is, the request is well-formed and ready to optimize.

It's the semantics that we're not sure about: Does this query compose functions in a way that makes sense? Did the user provide the correct arguments? That semantic analysis happens after the syntactic validation.

**Comment by smortensen** [comment]
Your intuition was right about the edge cases. Try this example:

```php
$query = 'insert(
	filter(
		slice,
		id = :id
	),
	{
		"id": :id
	}
)';
```

**Comment by smortensen** [comment]
I like how your "InsertDirectly" class can handle nested objects. That's excellent.

**Comment by choult** [comment]
@smortensen I've added another test case that (I believe) matches your edge case, and added validation for the input

**Comment by smortensen** [comment]
Here's the fully-parsed request that is troublesome:
[3,"insert",[[3,"filter",[[2,"slice"],[3,"equal",[[2,"id"],[1,"id"]]]]],[4,{"id":[1,"id"]}]]]

You can see that a property with an unusual name ("slice") is being treated as though it were the "slice" function. This property is not a function, but it's getting stripped from the request as though it were a function. PHP notices are being generated because of the unexpected structure of the token, and the end-result is a broken request.

But this is a well-formed query: There's nothing wrong with it all. It's syntactically and semantically correct. The only thing that's unusual about the request is one of its properties happens to have the same name as one of our functions. But that shouldn't be an issue for us: We should be able to handle properties with any name.

**Comment by smortensen** [comment]
Okay, I've written a unit test to show the input and output. Here's the commit:
https://github.com/datto/cinnabari/commit/86ee7ddb7480d41806ca04262624ff9c37994ecc

Anyway, I hope that helps clarify the expected behavior! I feel like you would have had no problem writing a perfect solution if the behavior had been clear from the beginning. I hope this helps.

**Comment by choult** [comment]
Thanks for the clarification @smortensen - code updated! I'll double check my local error reporting settings as well, to ensure I don't miss warnings and notices.
